Importing library & loading dataset

In [101]:
import pandas as pd
import numpy as np
df= pd.read_csv("../raw_data/customer_shopping_behavior.csv")

In [102]:
print(f"Initial shape: {df.shape}")

Initial shape: (10000, 26)


Data Inspection

In [103]:
print(df.head(50))

   Transaction ID Customer ID Purchase Date  Age  Gender       Location  \
0       TXN507808  CUST102534    2023-01-01   51    Male        Chennai   
1       TXN503830  CUST101256    2023-01-01   27  Female         Mumbai   
2       TXN501496  CUST100490    2023-01-01   57    Male      Hyderabad   
3       TXN501538  CUST100500    2023-01-01   22  Female           Pune   
4       TXN504265  CUST101399    2023-01-01   30    Male  Visakhapatnam   
5       TXN509250  CUST102952    2023-01-01   57    Male          Kochi   
6       TXN501814  CUST100596    2023-01-01   42    Male      Hyderabad   
7       TXN500031  CUST100007    2023-01-01   47    Male           Pune   
8       TXN501550  CUST100505    2023-01-01   37  Female      Ahmedabad   
9       TXN505122  CUST101684    2023-01-01   33    Male  Visakhapatnam   
10      TXN501568  CUST100510    2023-01-01   59    Male        Chennai   
11      TXN508080  CUST102612    2023-01-01   25    Male         Mumbai   
12      TXN501212  CUST10

In [104]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 26 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   Transaction ID          10000 non-null  object
 1   Customer ID             10000 non-null  object
 2   Purchase Date           10000 non-null  object
 3   Age                     10000 non-null  int64 
 4   Gender                  10000 non-null  object
 5   Location                10000 non-null  object
 6   Online/Offline          10000 non-null  object
 7   Online Store            7751 non-null   object
 8   Category                10000 non-null  object
 9   Item Purchased          10000 non-null  object
 10  Brand                   10000 non-null  object
 11  Color                   10000 non-null  object
 12  Size                    8965 non-null   object
 13  Quantity                10000 non-null  int64 
 14  Purchase Amount (₹)     10000 non-null  int64 
 15  Dis

In [105]:
df.describe()

,Age,Quantity,Purchase Amount (₹),Discount (%),Shipping Charge (₹),Delivery Time (Days),Review Rating,Previous Purchases
count,10000.000000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000
mean,38.278100,1.425100,1889.882000,32.916300,41.49210,3.703600,3.874900,3.211400
std,12.133041,0.761346,2102.301836,21.440617,64.30923,2.726223,0.912871,3.951313
min,18.000000,1.000000,67.000000,0.000000,0.00000,0.000000,1.000000,0.000000
25%,28.000000,1.000000,600.000000,12.000000,0.00000,1.000000,3.000000,0.000000
50%,38.000000,1.000000,1207.000000,33.000000,0.00000,4.000000,4.000000,2.000000
75%,49.000000,2.000000,2376.000000,52.000000,79.00000,6.000000,5.000000,4.000000
max,59.000000,4.000000,22306.000000,70.000000,199.00000,8.000000,5.000000,28.000000


In [106]:
df.duplicated().sum()

0

- Note:
    - Duplicates Not Found. 

#### Data to Clean:
- 1. Null Value Check and Handling.
- 2. Renaming Location column to Cities column.
- 3. Convert Purchase Date to relevant Dtype.

### 1. Null Value Check and Handling.

In [107]:
# Null Value Check:
df.isnull().sum().sort_values(ascending=False)

Festival/Sale             6788
Online Store              2249
Delivery Speed            2249
Size                      1035
Transaction ID               0
Purchase Amount (₹)          0
Previous Purchases           0
Return Status                0
Review Rating                0
Payment Method               0
Subscription Status          0
Delivery Time (Days)         0
Shipping Charge (₹)          0
Discount (%)                 0
Quantity                     0
Customer ID                  0
Color                        0
Brand                        0
Item Purchased               0
Category                     0
Online/Offline               0
Location                     0
Gender                       0
Age                          0
Purchase Date                0
Frequency of Purchases       0
dtype: int64

Note:
- Found nulls in Column:
    - A. Festival/Sale             6788
    - B. Online Store              2249
    - C. Delivery Speed            2249
    - D. Size                      1035

##### A. Null Value Handling for Column Festival/Sale

In [108]:
# 1. Column Unique Val check:
df['Festival/Sale'].unique()

array([nan, 'Republic Day Sale', 'Holi Sale', 'Amazon Great Summer Sale',
       'Independence Day Sale', 'Big Billion Days', 'Diwali Sale',
       'End of Season Sale'], dtype=object)

In [109]:
# 2. Null Val cross-relation check between 'Festival/Sale','Online Store':

df[df[['Festival/Sale','Online Store']].isnull().any(axis=1)][['Festival/Sale','Online Store']]

,Festival/Sale,Online Store
0,NaN,Flipkart
1,NaN,NaN
2,NaN,Flipkart
3,NaN,Myntra
4,NaN,Myntra
...,...,...
9987,End of Season Sale,NaN
9989,End of Season Sale,NaN
9991,End of Season Sale,NaN
9992,End of Season Sale,NaN


In [110]:
# 3. Null Value Handling:

df['Festival/Sale'] = df['Festival/Sale'].fillna('Regular Day')

In [111]:
# 4. Verifing Null Value Replacement:

print(df['Festival/Sale'].isna().sum())

0


##### B. Null Value Handling for Column Online Store:

In [112]:
# 1. Null Value Handling:

df.loc[
    (df['Online/Offline'] == 'Offline') & 
    (df['Online Store'].isna()),
    'Online Store'
] = 'In-Store Purchase'

# -- Note: Where the Online/Offline was 'Offline' &  the Online Store was missing, I replaced the missing Online Store with "In-Store Purchase"

In [113]:
# 2. Verifing Null Value Replacement:

df[['Online/Offline', 'Online Store']].head(10)

,Online/Offline,Online Store
0,Online,Flipkart
1,Offline,In-Store Purchase
2,Online,Flipkart
3,Online,Myntra
4,Online,Myntra
5,Offline,In-Store Purchase
6,Online,Myntra
7,Online,Myntra
8,Offline,In-Store Purchase
9,Online,Amazon


##### C. Null Value Handling for Column Delivery Speed

In [114]:
# C(1). Null Value Handling:
df.loc[
    (df['Delivery Time (Days)'] == 0) & 
    (df['Delivery Speed'].isna()), 
    'Delivery Speed'
] = 'N/A (Offline)'

## Note: Where the delivery time was 0 days &  the Delivery Speed was missing, I replaced the missing Delivery Speed with "N/A (Offline)"

In [115]:
# C(2). Verifing Null Value Replacement:

print(df['Delivery Speed'].isna().sum())

0


##### D. Null Value Handling for Column Size

In [116]:
# 1. Column Unique Val check:
df.loc[df['Size'].isna(), 'Category'].unique()

array(['Accessories'], dtype=object)

Note:
- All Null values are from the same category.

In [117]:
# 2. Null Val cross-relation check between Category, Item Purchased & Size:

df[df[['Category', 'Item Purchased','Size']].isnull().any(axis=1)][['Category', 'Item Purchased', 'Size']]

,Category,Item Purchased,Size
5,Accessories,Watch,NaN
33,Accessories,Watch,NaN
34,Accessories,Sunglasses,NaN
44,Accessories,Wallet,NaN
47,Accessories,Wallet,NaN
...,...,...,...
9935,Accessories,Belt,NaN
9954,Accessories,Watch,NaN
9974,Accessories,Watch,NaN
9988,Accessories,Sunglasses,NaN


Note:
- No specific sub-category relation found for the nulls.
- Thus, filling up with Unknown

In [118]:
# 3. Null value handling:

df['Size'] = df['Size'].fillna('Unknown')

In [119]:
# 4.  Verifying Null Value Replacement:

print(df['Size'].isna().sum())

0


##### B. Renaming the Location column with Cities column.

In [120]:
df.rename(columns={'Location': 'Cities'}, inplace=True)

##### C. Convert Purchase Date to relevant Dtype

In [121]:
df['Purchase Date'] = pd.to_datetime(df['Purchase Date'], errors = 'coerce')

print(df['Purchase Date'].dtype)

datetime64[ns]


# Feature Engineering:
- A. Creating a New Region Column.
- B. Creating column for footwear and clothing size. 
- C. Extracting Parts of date.
- D. Creating Age group 

##### A. Creating a New Region Column:

In [122]:
# 1. All unique city names:

df['Cities'].unique()

array(['Chennai', 'Mumbai', 'Hyderabad', 'Pune', 'Visakhapatnam', 'Kochi',
       'Ahmedabad', 'Chandigarh', 'Delhi', 'Bangalore', 'Indore',
       'Jaipur', 'Kolkata', 'Patna', 'Lucknow', 'Nagpur', 'Noida',
       'Surat', 'Bhubaneswar', 'Gurgaon'], dtype=object)

In [123]:
# 2. Mapping the cities

region_map = {
    'Delhi': 'North',
    'Noida': 'North',
    'Gurgaon': 'North',
    'Chandigarh': 'North',
    'Jaipur': 'North',
    'Lucknow': 'North',

    'Chennai': 'South',
    'Hyderabad': 'South',
    'Visakhapatnam': 'South',
    'Kochi': 'South',
    'Bangalore': 'South',

    'Mumbai': 'West',
    'Pune': 'West',
    'Ahmedabad': 'West',
    'Nagpur': 'West',
    'Surat': 'West',

    'Kolkata': 'East',
    'Patna': 'East',
    'Bhubaneswar': 'East',

    'Indore': 'Central'
}

df['Region'] = df['Cities'].map(region_map)

In [ ]:
# 3. Popping out the region column:

region = df.pop('Region')

# Finding the position of the Cities column:

city_loc = df.columns.get_loc('Cities')

# Inserting Region column immediately after Cities column:

df.insert(city_loc + 1, 'Region', region)


##### B. Creating column for Footwear & Clothing Size. 

In [125]:
# 1. Category column unique vals:
df['Category'].unique()

array(['Clothing', 'Accessories', 'Footwear'], dtype=object)

In [126]:
# 2. Footwear Size:
df['Footwear_Size'] = np.where(df['Category'] == 'Footwear', df['Size'], 'Not Applicable')

# 3. Clothing Size:
df['Clothing_Size'] = np.where(df['Category'] != 'Footwear', df['Size'], 'Not Applicable')

In [ ]:
# 4. Finding the position of the Size column:

size_loc = df.columns.get_loc('Size')

# 5. Inserting Clothing_Size column immediately after Size column:

df.insert(size_loc + 1, 'Clothing_Size', Clothing_Size)

# 6. Inserting Footwear_Size column immediately after Size column:

df.insert(size_loc + 2, 'Footwear_Size', Footwear_Size)

In [ ]:
# 7. Dropping Size column:
df.drop( columns = ['Size'], inplace = True)

In [ ]:
# 8. Column after changes:
df[['Category', 'Footwear_Size', 'Clothing_Size']].head(10)

,Category,Footwear_Size,Clothing_Size
0,Clothing,Not Applicable,M
1,Clothing,Not Applicable,L
2,Clothing,Not Applicable,XXL
3,Clothing,Not Applicable,XS
4,Accessories,Not Applicable,Free Size
5,Accessories,Not Applicable,Unknown
6,Footwear,10,Not Applicable
7,Clothing,Not Applicable,S
8,Clothing,Not Applicable,M
9,Clothing,Not Applicable,XS


##### C. Extracting Parts of Date

In [ ]:
# 1. Finding the position of the Purchase Date column:
purchase_date_loc = df.columns.get_loc('Purchase Date')


# 2. Extracting Year and inserting it immediately after the Purchase Date column:
df.insert(purchase_date_loc + 1, 'Year', df['Purchase Date'].dt.year)


# 3. Extracting Month and inserting it immediately after the Year column:
df.insert(purchase_date_loc + 2, 'Month', df['Purchase Date'].dt.month)


# 4. Extracting Day and inserting it immediately after the Month column:
df.insert(purchase_date_loc + 3, 'Day', df['Purchase Date'].dt.day)


# 5. Extracting Weekday Name and inserting it immediately after the Day column:
df.insert(purchase_date_loc + 4, 'Weekday', df['Purchase Date'].dt.day_name())


D. Creating Age group:

In [ ]:
def age_group(age):
    if age < 18:
        return 'Under 18'
    elif age <= 25:
        return '18-25'
    elif age <= 35:
        return '26-35'
    elif age <= 45:
        return '36-45'
    elif age <= 60:
        return '46-60'
    else:
        return 'Above 60'
    
# Finding the position of the age column:
age_loc = df.columns.get_loc('Age')

# Inserting the Age Group column immediately after the age column:
df.insert(age_loc + 1, 'Age Group', df['Age'].apply(age_group))

In [ ]:
df.head()

,Transaction ID,Customer ID,Purchase Date,Year,Month,Day,Weekday,Age,Age Group,Gender,...,Delivery Speed,Delivery Time (Days),Subscription Status,Payment Method,Review Rating,Return Status,Previous Purchases,Frequency of Purchases,Footwear_Size,Clothing_Size
0,TXN507808,CUST102534,2023-01-01,2023,1,1,Sunday,51,46-60,Male,...,Same Day,1,Yes,Cash on Delivery,4,Not Returned,0,Fortnightly,Not Applicable,M
1,TXN503830,CUST101256,2023-01-01,2023,1,1,Sunday,27,26-35,Female,...,N/A (Offline),0,No,UPI,5,Not Returned,0,Fortnightly,Not Applicable,L
2,TXN501496,CUST100490,2023-01-01,2023,1,1,Sunday,57,46-60,Male,...,Standard,7,Yes,UPI,3,Not Returned,0,Rarely,Not Applicable,XXL
3,TXN501538,CUST100500,2023-01-01,2023,1,1,Sunday,22,18-25,Female,...,Standard,6,Yes,Net Banking,4,Not Returned,0,Fortnightly,Not Applicable,XS
4,TXN504265,CUST101399,2023-01-01,2023,1,1,Sunday,30,26-35,Male,...,Express,2,No,Debit Card,4,Not Returned,0,Weekly,Not Applicable,Free Size


In [132]:
print(f"Shape After Preprocessing:", {df.shape})

Shape After Preprocessing: {(10000, 33)}


In [133]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 33 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Transaction ID          10000 non-null  object        
 1   Customer ID             10000 non-null  object        
 2   Purchase Date           10000 non-null  datetime64[ns]
 3   Year                    10000 non-null  int32         
 4   Month                   10000 non-null  int32         
 5   Day                     10000 non-null  int32         
 6   Weekday                 10000 non-null  object        
 7   Age                     10000 non-null  int64         
 8   Age Group               10000 non-null  object        
 9   Gender                  10000 non-null  object        
 10  Cities                  10000 non-null  object        
 11  Region                  10000 non-null  object        
 12  Online/Offline          10000 non-null  object 

Exporting Clean Dataset

In [134]:
df.to_csv("../clean_data/clean_customer_shopping_behavior_india.csv", index=False)

## Conclusion

- Performed an initial dataset inspection of the dataset using `.info()` and `.describe()`.

- Conducted duplicate record analysis.

- Identified a datatype inconsistency in the `Purchase Date` column.

- Renamed `Location` column to `Cities` to improve dataframe readablity.

- Identified and handled missing values in the following columns:
  - `Festival/Sale`
  - `Online Store`
  - `Delivery Speed`
  - `Size`

- Corrected datatype inconsistencies by converting the `Purchase Date` column from `object` to `datetime`.

- Engineered new business-relevant features, including:
  - Created a `Region` column immedietly after `Cities` using a mapping funtion.
  - Created a `Footwear_Size` & `Clothing_Size` column immedietly after `Size` then dropped the `Size` column.
  - Extracted `Year`, `Month`, `Day` and `Weekday` from the `Purchase Date` column and inserted them immedietly after it.
  - Created a `Age Group` column immedietly after `Age` using a user-defined (`def`) funtion

- Verified the dataset's shape and confirmed successful handling of missing values after preprocessing.

- Exported the cleaned dataset for further analysis.